### PR TITLE
Error when `POSTGRES_PASSWORD` is unset like mysql

### LINKED ISSUE #580: POSTGRES_PASSWORD changes authentication settings unexpectedly
I believe that current authentication configuration in `docker-entrypoint.sh` is less secure than it should be.  I have a problem with `$POSTGRES_PASSWORD` changing the authentication settings and not adequately warning user about it.

Currently if the password is not supplied `host all all all trust` is added to `pg_hba.conf` , disabling the password checks for all users. I think this goes way too far:

1) I honestly believe that there should be no way for the line `host all all all trust` to appear in configuration unless user explicitly sets `ACCEPT_ANY_PASSWORD_FOR_ANY_USER=true`
2) Currently any mistake in configuration (for example misspelling  `POSTGRES_PASSWORD`) results in disabled authentication
3) Disabled authentication is completely silent. User could believe that password authentication is working, since the DB accepts the password. 
4) Warning displayed if `POSTGRES_PASSWORD` wasn't set does not convey that password checks were disabled altogether. 
> No password has been set for the database. This will allow anyone with access to the Postgres port to access your database.

Sounds like you can fix it by setting the password, but the password would be just ignored.

I had an unpleasant experience of discovering my development DB for a personal project consuming 100% of server's CPU because someone was mining some kind of crypto coin on it.
Initially I was using `POSTGRES_PASSWORD`, but later I decided to move password configuration for "postgres" user together with creating additional, less privileged users into an init SQL file in `/docker-entrypoint-initdb.d/` (amazing feature, btw). As you might imagine, I did not believe that this would disable all password checks altogether, and since my server continued to grant access with my password I had no idea that I was completely insecure.

I did not expect (and didn't need) high security from DB with an exposed port and without SSL, but having literally no security is a really strange decision.

I've read issues about security of this image (#3 #29 #31), and it seems like "security is annoying and not our problem" argument is winning. I'd much rather have official images be more secure by default, but I feel like this fight is futile. So here are my suggestions, from less bothersome to more secure:

1) Change the warning to clearly state "all authentication is disabled, any password is valid for any user, change pg_hba.conf or set $POSTGRES_PASSWORD if this behavior is undesirable" or something like that. Add this warning to readme.
2) If `POSTGRES_PASSWORD` is not set (better yet, if `ALLOW_EMPTY_PASSWORD` is set), then set it to an empty string and keep the `md5` authentication always enabled. That way developers can still easily connect to a DB with empty password, but changing the password would never be silently ignored, like it was in my case.
3) If `POSTGRES_PASSWORD` is not set (better yet, if `GENERATE_RANDOM_PASSWORD` is set), then set password to a random string and print it on initial launch.

Official MySQL image allows to get second or third behavior by setting `MYSQL_ALLOW_EMPTY_PASSWORD` or `MYSQL_RANDOM_ROOT_PASSWORD` environment variables. Alternatively, you can set `MYSQL_ROOT_PASSWORD` directly. One of the three variables must be set explicitly, or MySQL container fails to launch.

For backwards compatibility `ACCEPT_ANY_PASSWORD_FOR_ANY_USER` or similarly explicitly named environment variable could be added. If it is set to true then `host all all all trust` is added to the config.

I'm willing to write the necessary code, but I'd like to know beforehand which approach would be accepted. IMHO MySQL's approach is the correct one, it is both flexible and explicit, but it's up to the maintainers.

### PR DESCRIPTION
**Pull request by yosifkit**
Add `POSTGRES_HOST_AUTH_METHOD` to bring back old behavior and be similar to `MYSQL_ALLOW_EMPTY_PASSWORD`, but add warning when "trust" is used since it disables all passwords

Fixes #580.

### CODE REVIEW

File: docker-entrypoint.sh [hunk 1]
```diff
@@ -103,22 +105,31 @@ docker_verify_minimum_env() {
...
+	if [ -n "$POSTGRES_DISABLE_PASSWORDS" ]; then
 		cat >&2 <<-'EOWARN'
 			****************************************************
-			WARNING: No password has been set for the database.
+			WARNING: POSTGRES_DISABLE_PASSWORDS has been set.
 			         This will allow anyone with access to the
-			         Postgres port to access your database. In
-			         Docker's default configuration, this is
+			         Postgres port to access your database without
+			         a password, even if POSTGRES_PASSWORD is set.
```
Comment (yosifkit) [inline docker-entrypoint.sh]: Perhaps this (or the `pg_setup_hba_conf` function) should also print what is added to `pg_hba.conf`?

Comment (tianon) [reply docker-entrypoint.sh]: Maybe just a link to https://www.postgresql.org/docs/12/auth-trust.html ?

**Comment by Andrew-Morozko** [comment]
Thank you for considering my request seriously. I'm much happier with this behaviour.

One little nitpick is the result when both `POSTGRES_PASSWORD` and `POSTGRES_DISABLE_PASSWORDS` are set. Currently it ignores the supplied `POSTGRES_PASSWORD`. If we're going for clarity this should error-out, otherwise `POSTGRES_PASSWORD` should overrule the `POSTGRES_DISABLE_PASSWORDS`.

Ether way, I'm much less sympathetic if someone forgets to remove `POSTGRES_DISABLE_PASSWORDS` from production config, but still, *somebody* will make that mistake given time...

### Proposed patches 
Add to `docker_verify_minimum_env`, before `if [ -n "$POSTGRES_DISABLE_PASSWORDS" ]`

Error-out on incompatible options:
```bash
	if [ -n "$POSTGRES_PASSWORD" ] && [ -n "$POSTGRES_DISABLE_PASSWORDS" ]; then
		cat >&2 <<-'EOE'
			Error: Both POSTGRES_PASSWORD and POSTGRES_DISABLE_PASSWORDS are specified.
			       You must specify one and only one.
		EOE
		exit 1
	fi
```

**OR**

Ignore `POSTGRES_DISABLE_PASSWORDS` on incompatible options:
```bash
	if [ -n "$POSTGRES_PASSWORD" ] && [ -n "$POSTGRES_DISABLE_PASSWORDS" ]; then
		cat >&2 <<-'EOE'
			WARNING: Both POSTGRES_PASSWORD and POSTGRES_DISABLE_PASSWORDS are specified.
			         Using POSTGRES_PASSWORD, ignoring POSTGRES_DISABLE_PASSWORDS.
		EOE
		unset POSTGRES_DISABLE_PASSWORDS
	fi
```

P.S.: Sorry for slow feedback, end of the year crazyness left me with full inbox...

**Comment by ringerc** [comment]
Erroring on incompatible options probably makes more sense.

The error wording looks sensible.

I'm surprised you had to change so many places. No include files or snippets? Seems like plenty of room to miss one and leave things inconsistent. But that's not really the issue at hand.

Thankyou for this. I hope you're able to get it merged, it'd help reduce the rate at which potentially wormable wide-open postgres instances are created by people who don't realise what they're doing.

**Comment by tianon** [comment]
> Erroring on incompatible options probably makes more sense.

Sure, but in this case they aren't actually incompatible -- they're orthogonal.  When one sets `POSTGRES_DISABLE_PASSWORDS`, the database is configured to allow all connections, regardless of their authentication parameters.  Separately, `POSTGRES_PASSWORD` configures the password which is set on the default user, and that gets set/stored regardless of the authentication method.

In other words, if the user asks us to disable password authentication, we do so (including spitting out a large warning).

Rethinking in this context, I think the name of the variable is misleading and should more directly correspond to what it does instead of what the user intends -- we're not actually disabling "passwords" in PostgreSQL but rather asking PostgreSQL not to prompt for them by setting the authentication method to be `trust`.

So, perhaps we should instead call it something more explicit like `POSTGRES_AUTH_METHOD` and warn when the value is set to `trust`.  This would have the added benefit of allowing users to opt into newer authentication methods like `scram-sha-256` (https://www.postgresql.org/docs/11/auth-password.html), increasing security further.

> I'm surprised you had to change so many places. No include files or snippets? Seems like plenty of room to miss one and leave things inconsistent. But that's not really the issue at hand.

They're all templated -- we just have to commit the templating result so that Travis tests the right thing.

**Comment by yosifkit** [comment]
Swapped `POSTGRES_DISABLE_PASSWORDS` to `POSTGRES_HOST_AUTH_METHOD` so that users configure the value of the `auth-method` rather than an opt-in to "disable passwords".

I'll try to get a docs PR going soon that will link to https://www.postgresql.org/docs/current/auth-pg-hba-conf.html

**Comment by yosifkit** [comment]
Docs PR created: https://github.com/docker-library/docs/pull/1653

**Comment by BYK** [comment]
>This is a breaking change, but I think it's one that makes sense overall (and that the benefit outweighs the pain in this instance).

Burnt by this @tianon - now our setup requires `POSTGRES_HOST_AUTH_METHOD=trust` to be passed. Not sure how this passed the bar for a "patch" release.

**Comment by levicole** [comment]
I agree with @BYK. We can't be the only two people that were burnt by this. Not appropriate for a patch release at all.

**Comment by tianon** [comment]
We don't have an "image version" -- our published image version is the version of PostgreSQL itself.  The fact that this went out with a patch release is entirely coincidental, and it would've gone out as an update to the existing images regardless.  See https://github.com/docker-library/postgres/issues/681#issuecomment-586517154 for further explanation.

The workaround for folks who want to stay insecure is to set `POSTGRES_HOST_AUTH_METHOD=trust` or if you **know** you cannot tolerate image updates, pin to an explicit content-addressable digest (which is a good practice to follow regardless).

**Comment by BYK** [comment]
> The fact that this went out with a patch release is entirely coincidental, and it would've gone out as an update to the existing images regardless.

This is even more scarier and a shortcoming of the official images approach I guess?

>The workaround for folks who want to stay insecure is to set POSTGRES_HOST_AUTH_METHOD=trust

If the network is limited properly this is not insecure. Having good and more secure defaults is great but if you had a bad default you don't "fix" it in a minor version bump.

>or if you know you cannot tolerate image updates, pin to an explicit content-addressable digest (which is a good practice to follow regardless).

See above. There's a reason why semver exists and people rely on those semantics. We can tolerate changes in images as long as they are patch releases which was what we have set and here we are.

**Comment by BYK** [comment]
> The fact that this went out with a patch release is entirely coincidental, and it would've gone out as an update to the existing images regardless.

This is even more scarier and a shortcoming of the official images approach I guess?

>The workaround for folks who want to stay insecure is to set POSTGRES_HOST_AUTH_METHOD=trust

If the network is limited properly this is not insecure. Having good and more secure defaults is great but if you had a bad default you don't "fix" it in a minor version bump.

>or if you know you cannot tolerate image updates, pin to an explicit content-addressable digest (which is a good practice to follow regardless).

See above. There's a reason why semver exists and people rely on those semantics. We can tolerate changes in images as long as they are patch releases which was what we have set and here we are.

**Comment by tianon** [comment]
None of the official images follow semver (that I'm aware of).

**Comment by tierra** [comment]
I do think it's a good change, for good reasons, but like @BYK , I would have expected the change to apply to all *new* PostgreSQL major version image tags. Given the timing here, that would have just been in `postgres:13` and later. Some middle ground might have been fine too though: Impose some pain by retroactively applying it to `postgres:12` (and in all future major tags), but still leave `postgres:11` and below alone.

**Comment by EliotBerriot** [comment]
Been hit by that today, wanted to thank the maintainers for their work. This is the first time I've had a breaking change with this image, and eventhough it's annoying, the warning makes it rather easy to understand what's going on.

As far as I understand, the versionning (11.7, 12, etc.) applies to the version of postgres that is shipped in the image, not the image API/configuration itself.

**Comment by mltsy** [comment]
I also didn't realize these official-images images were essentially versionless.  I'll look for advice on that in the main repo docs, but I guess it sounds like pinning to a specific digest is the recommended solution?  Thanks!  (It would be great to add a note to this effect in the postgres image README/docs, even if it is documented somewhere in the official-images repository usage instructions...).  And I agree about the usefulness of a deprecation warning for breaking changes in that case, but also understand that means more resources, so advice in the README about pinning a specific digest would be a good substitute! 🙂

**Comment by sgerrand** [comment]
> There's a reason why semver exists and people rely on those semantics. We can tolerate changes in images as long as they are patch releases which was what we have set and here we are.

FWIW, consumers of these images who want predictable content can use a [digest reference](https://docs.docker.com/registry/spec/api/#digest-parameter#content-digests). I too was burnt but have been happy to apply the configuration changes as a tradeoff for a potentially nasty incident.

**Pull request merged**